### DISCUSSION TITLE
Do not implement Copy for types with "&mut self" methods

### DISCUSSION THREAD

**Discussion by newpavlov** [rust-lang-nursery/api-guidelines#116] [API Guidelines]
Implicit copies of types with methods accepting `&mut self` can produce surprising results for library users. Thus it's probably worth to add recommendation to avoid deriving `Copy` trait for such types.

Related [issue](https://github.com/RustCrypto/hashes/issues/30).

**Comment by dtolnay** [comment]
Thanks! We definitely need a guideline for this.

Relevant discussions in the standard library related to Copy for range iterators in the following sequence of PRs:

- [remove Copy impls from remaining iterators](https://github.com/rust-lang/rust/pull/21809) (merged)
- [Revert "remove Copy impls from remaining iterators"](https://github.com/rust-lang/rust/pull/21846) (closed)
- [Derive Copy for std::ops::Range and RangeFrom](https://github.com/rust-lang/rust/pull/27186) (closed)

One of the motivating examples there was:

```rust
let mut iter = 0..n;
for i in iter { if i > 2 { break; } }
iter.collect()
```

With a Copy impl, this iterates over a copy of the range and then `collect` gets all 0..n. Without a Copy impl, the compiler catches this and makes you use `clone()` to clone the range or `for i in &mut iter` to consume elements from the range.
